doc/architecture/decisions/0008-add-first-class-citizen-fields-to-applications.md

```diff
@@ -0,0 +1,139 @@
+# 8. Add First-Class Citizen Fields to Applications
+
+Date: 2022-10-31
+
+## Status
+
+Accepted
+
+## Context
+
+Approved Premises applications are currently stored in the database as a JSON
+object. This allows the format of the questionnaire to be flexible, simple, and
+take into account any changes in the questionnaire format, whilst also
+preserving applications that have been submitted in an older format.
+
+However, when processing an application, there are some situations where we need
+easy access to particular fields within the JSON application object. These include
+(but are not limited to):
+
+- If an application is for a women-only AP
+- If an application is for a PIPE placement
+- The target location
+- The desired start date
+
+These are important fields that need to be queried on, and made into first class
+citizens within the database table, without sacrificing flexibility.
+
+## Decision
+
+At the point of submission of a completed application, we will use [JsonLogic](https://jsonlogic.com)
```

Comment:
So would `submittedAt` remain a top level property that the frontend sets on the PUT request to trigger a submission - or we would use a JSON Logic query each time a PUT is made to pull out a submittedAt from `data` then continue if it's not null?

Reply:
This would only be done once when the final application is submitted. I guess we'll need to have a new endpoint that submits the final application

Reply:
yes, "submission" seems like an important milestone -- it's not just another type of persistence, it:

- generates the permanent `document`
- prevents further changes to the application
- advances the application to the assessor allocation stage

so an endpoint like `POST applications/{id}/submission` seems appropriate.